### ISSUE TITLE
Improve compatibility with CI/CD

### ISSUE THREAD

**Issue by matzeeable** [lerna/lerna#2346]
## Context
I have setup `lerna` and I want to use `lerna version`/ `lerna publish` within my CI/CD pipeline in GitLab.

## Expected Behavior
Get branch name correctly.

## Current Behavior
In GitLab the branch is checked out as detached HEAD state so `currentBranch` is always `HEAD` which leads to issues:

![](https://i.imgur.com/HyLLdZg.png)

## Possible Solution

[semantic-release](https://github.com/semantic-release/semantic-release) is using [`env-ci`](https://github.com/pvdlg/env-ci) to determine the correct branch name: https://github.com/semantic-release/semantic-release/blob/7a939a897097a730f1579002ed641a583d3381c4/index.js#L6-L26

`lerna` is using an own `git` command instead of using `env-ci`:

https://github.com/lerna/lerna/blob/5344820fc65da081d17f7fd2adb50ffe7101905b/commands/version/lib/get-current-branch.js#L11

I am open to create another PR (beside my others https://github.com/lerna/lerna/issues/2333) for this solution.

**A solution for this is to use `env-ci` additionally, and use the `git` command if the returned `branch` is `undefined`**

Apart from that I found another code where `is-ci` is used:

https://github.com/lerna/lerna/blob/5344820fc65da081d17f7fd2adb50ffe7101905b/core/command/index.js#L129

Perhaps it should be considered to remove that package and also replace it with `env-ci`. But there are breaking changes because both packages support different CI systems:

| env-ci                      | is-ci (ci-info)      |
|-----------------------------|----------------------|
| AWS CodeBuild               | AWS CodeBuild        |
| AppVeyor                    | AppVeyor             |
|                             | Azure Pipelines      |
| Bamboo                      | Bambooby Atlassian   |
| Bitbucket                   | Bitbucket Pipelines  |
| Bitrise                     | Bitrise              |
| Buddy                       | Buddy                |
| Buildkite                   | Buildkite            |
| CircleCI                    | CircleCI             |
| Cirrus CI                   | Cirrus CI            |
| Codeship                    | Codeship             |
| Codefresh                   |                      |
| Drone                       | Drone                |
|                             | dsari                |
| GitHub Actions              |                      |
| GitLab CI/CD                | GitLab CI            |
|                             | GoCD                 |
|                             | Hudson               |
| Jenkins                     | Jenkins CI           |
|                             | Magnum CI            |
|                             | Netlify CI           |
|                             | Nevercode            |
| Puppet                      |                      |
| Sail CI                     | Sail CI              |
| Scrutinizer                 |                      |
| Semaphore                   | Semaphore            |
| Shippable                   | Shippable            |
|                             | Solano CI            |
|                             | Strider CD           |
|                             | TaskCluster          |
| TeamCity                    | TeamCityby JetBrains |
| Travis CI                   | Travis CI            |
| Visual Studio Team Services |                      |
| Wercker                     |                      |

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.18.4 |
| `npm --version`  | 6.4.1 |
| `yarn --version` | 1.19.1 |
| `node --version` | v10.15.3 |

| OS | Version |
| --- | --- |
| WSL | On Windows 10 |

**Comment by evocateur** [comment]
`env-ci` [supports](https://github.com/pvdlg/env-ci#supported-ci) enough `isCi` that it seems reasonable to back-fill when `branch` is `undefined`.

The internal logic of `env-ci`'s equivalent to our `getCurrentBranch` is [interesting](https://github.com/pvdlg/env-ci/blob/59bb812f51f3d6a518f5f8aedf9eb74cc1bf6782/lib/git.js#L13-L24):
```js
const headRef = execa.sync("git", ["rev-parse", "--abbrev-ref", "HEAD"], options).stdout;

if (headRef === "HEAD") {
  const branch = execa
    .sync("git", ["show", "-s", "--pretty=%d", "HEAD"], options)
    .stdout.replace(/^\(|\)$/g, "")
    .split(", ")
    .find(branch => branch.startsWith("origin/"));

  return branch ? branch.match(/^origin\/(.+)/)[1] : undefined;
}

return headRef;
```

I suppose it would technically be a breaking change to switch `is-ci` entirely to `env-ci`, yes. I hope to have time to make a breaking release soon (dropping node < 10, etc), so it wouldn't hurt to sketch out.

**Comment by matzeeable** [comment]
Ok. I need this functionality soon. Can we go without `env-ci` and use the coding you suggested? If you want I can start a PR.

**Comment by evocateur** [comment]
I'm not even sure that code would work when `lerna version` tries to create a commit and push it back to the remote. You're still in a detached `HEAD`, and in my experience that needs to be dealt with _before_ calling `lerna publish`.

**Comment by matzeeable** [comment]
Yeah the `version` command needs to be adjusted. I used `semantic-release` alot and there the CI/CD integration worked pretty well. I looked at their coding and the also push "over" the detached `HEAD` state:

https://github.com/semantic-release/git/blob/16edc0b7794482c3f57e144c3f4cf01abe0b3205/lib/git.js#L1

Also checkout https://stackoverflow.com/a/40929378/5506547 and https://stackoverflow.com/a/52338714/5506547

What do you think?

**Comment by matzeeable** [comment]
Hi @evocateur ! Sorry for the ping but we plan to release our major release of our OS WordPress project. I need to get `lerna` work on CI - if you want I can start a PR with following changes:

- detect branch like `semantic-release` does with fallback to current implementation
- push back to branch like `semantic-release` does

**Comment by JounQin** [comment]
https://github.com/mdx-js/mdx/pull/862#issuecomment-565804348

It seems lerna breaks on CI after `3.18.1`?

**Comment by JounQin** [comment]
After downgrading `lerna` to `3.17.0`, it seems working.

https://circleci.com/gh/mdx-js/mdx/1591?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by matzeeable** [comment]
Another issue on GitLab CI, that `postversion` lifecycle could not run:

```
 lerna WARN lifecycle @wp-reactjs-multi-starter/wp-reactjs-starter@1.2.0~version: cannot run in wd @wp-reactjs-multi-starter/wp-reactjs-starter@1.2.0 yarn grunt postversion (wd=/builds/devowl.io/wp-reactjs-starter/plugins/wp-reactjs-starter)
```

Has anyone experienced this before?

**Comment by jdalrymple** [comment]
Im getting similar problems within [Gitlab CI ](https://gitlab.com/jdalrymple/gitbeaker/-/jobs/519950116#L1649). Specifically `git merge-base --is-ancestor`

**Comment by jenniferpham** [comment]
I think you may need to checkout to master branch.  
I was able to get my CI to work and publish through Github Actions.  Even if you're not using Github Actions, maybe this could give you an idea of what to do.  I added SSH-key and Github token which had write access to the repo.

```
    steps:
      - name: Copy repo code
        uses: actions/checkout@v2
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          ssh-key: ${{ secrets.SSH_PRIVATE_KEY }}

     - name: Configure git credentials
        uses: OleksiyRudenko/gha-git-credentials@v1
        with:
          token: ${{ secrets.GITHUB_TOKEN }}

      - name: Publish New Version and Release
        env:
          NODE_AUTH_TOKEN: ${{secrets.NPM_TOKEN}}
          GH_TOKEN: ${{secrets.GITHUB_TOKEN}}
        run: |
          lerna publish --yes
```
